File: 刷题/cpp/792. 匹配子序列的单词数.cpp
//法一：分桶
class Solution {
public:
    int numMatchingSubseq(string s, vector<string>& words) {
        vector<queue<string>> d(26);
        for(auto& w:words){
            d[w[0]-'a'].emplace(w);
        }
        int ans=0;
        for(auto c:s){
            auto&& q=d[c-'a'];
            int size=q.size();
            while(size--){
                string t=q.front();
                q.pop();
                if(t.length()==1) ans++;
                else d[t[1]-'a'].emplace(t.substr(1));
            }
        }
        return ans;
    }
};
//法二：二分查找
class Solution {
public:
    int numMatchingSubseq(string s, vector<string>& words) {
        vector<vector<int>> pos(26);
        for(int i=0;i<s.length();i++){
            pos[s[i]-'a'].emplace_back(i);
        }
        int ans=words.size();
        for(auto str:words){
            if(str.length()>s.length()){
                ans--;
                continue;
            }
            int p=-1;
            for(char c:str){
                auto &ps=pos[c-'a'];
                auto it=upper_bound(ps.begin(),ps.end(),p);
                if(it==ps.end()){
                    ans--;
                    break;
                }
                p=*it;
            }
        }
        return ans;
    }
};

File: 刷题/golang/792. 匹配子序列的单词数.go
//法一：二分查找
func numMatchingSubseq(s string, words []string) int {
    pos:=[26][]int{}
    for i,c:=range s{
        pos[c-'a']=append(pos[c-'a'],i)
    }
    ans:=len(words)
    for _,w:=range words{
        if len(w) > len(s){
            ans--
            continue
        }
        p:=-1
        for _,c:=range w{
            ps:=pos[c-'a']
            j:=sort.SearchInts(ps,p+1)
            if j==len(ps){
                ans--
                break
            }
            p=ps[j]
        }

    }
    return ans
}
//法二：分桶
func numMatchingSubseq(s string, words []string) (ans int) {
    type pair struct{i,j int}
    ps :=[26][]pair{}
    for i,w:=range words{
        ps[w[0]-'a']=append(ps[w[0]-'a'],pair{i,0})
    }
    for _,c:=range s{
        q:=ps[c-'a']
        ps[c-'a']=nil
        for _,p:=range q{
            p.j++
            if p.j==len(words[p.i]){
                ans++
            }else{
                w:=words[p.i][p.j]-'a'
                ps[w]=append(ps[w],p)
            }
        }
    }
    return
}